mitodl/ocw-studio
# backpopulate archive videos

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1739

#### What's this PR do?
This PR adds a new management command, `backpopulate_archive_videos`, along with a batchable task behind it that does the actual work involved. The management command takes all `Website` objects, optionally filtered with `--filter` and does the following:

 - For each `Website`, find `WebsiteContent` objects where `metadata.video_files.archive_url` is not null
 - For each `WebsiteContent` object, find the associated video in the bucket specified by the `--bucket` argument and optionally the `--prefix` argument based on the path portion of `metadata.video_files.archive_url`
 - Sync the video we found to the buckets stored in these settings:
   - `AWS_STORAGE_BUCKET_NAME`
   - `AWS_PREVIEW_BUCKET_NAME`
   - `AWS_PUBLISH_BUCKET_NAME`
   - `AWS_OFFLINE_PREVIEW_BUCKET_NAME`
   - `AWS_OFFLINE_PUBLISH_BUCKET_NAME`

This work is being done as part of https://github.com/mitodl/hq/discussions/1197 and the reason we are moving the videos into all these different buckets is to truly import them into the `ocw-studio` system, as if the videos had been uploaded by content authors. Putting them into the storage and preview / publish buckets prevents a large sync from having to happen on each site after this command is run for the first time, as all the videos will already be in place. This allows us to deliver videos from legacy sites in mirror drives without having to conditionally pull the videos from a different bucket at build time depending on if the sites is considered "legacy" or not.

#### How should this be manually tested?
 - Make sure you have the following prerequisites set up (if unsure, check the readme):
   - Recent DB restore from production or RC with sync states and gdrive folders reset
   - Local Concourse integration configured and running
   - Local Minio integration configured and running
   - Theme assets pipeline pushed up to Concourse and run at least once
   - In your `.env`, these are the settings I used minus the minio user / password and aws access keys:
```
AWS_STORAGE_BUCKET_NAME=ol-ocw-studio-app
AWS_PREVIEW_BUCKET_NAME=ocw-content-draft
AWS_PUBLISH_BUCKET_NAME=ocw-content-live
AWS_ARTIFACTS_BUCKET_NAME=ol-eng-artifacts
RESOURCE_BASE_URL_DRAFT=https://draft.ocw.mit.edu
RESOURCE_BASE_URL_LIVE=https://live.ocw.mit.edu
STATIC_API_BASE_URL=https://live.ocw.mit.edu
OCW_HUGO_THEMES_BRANCH=main
OCW_HUGO_PROJECTS_BRANCH=main
```
 - Create a folder somewhere on your computer and open a terminal inside it
 - Run `aws s3 sync s3://ocw-website-mirror/OcwExport/InternetArchive/MIT6.811F14/ ./`
 - Log into your local Minio instance at http://localhost:9001 and:
   - Create a new bucket called `ocw-website-mirror`
   - In that bucket, create a folder called `OcwExport`, and in that folder create another folder called `InternetArchive`
   - In that folder create a folder called `MIT6.811F14`, then drag and drop the files we downloaded with the `s3 sync` command above into it
 - In `ocw-studio`, run `docker compose exec web ./manage.py backpopulate_archive_videos --bucket ocw-website-mirror --prefix OcwExport/InternetArchive --filter 6-811-principles-and-practice-of-assistive-technology-fall-2014`
 - After the command completes, log back into the Minio web interface at http://localhost:9001
 - Check the storage, preview and publish buckets listed above and verify that the videos have been copied into the appropriate places
 - Browse to http://localhost:8043/sites/6-811-principles-and-practice-of-assistive-technology-fall-2014/ and publish the site live
 - Back in Minio, verify in the preview and publish buckets that the videos have not been removed

#### Where should the reviewer start?
`content_sync/tasks.py/backpopulate_archive_videos_batch`
